alvr/server/cpp/platform/linux/FrameRender.cpp

```diff
@@ -0,0 +1,758 @@
+
+    VkImportSemaphoreFdInfoKHR impSemInfo = {};
+    impSemInfo.sType = VK_STRUCTURE_TYPE_IMPORT_SEMAPHORE_FD_INFO_KHR;
+    impSemInfo.semaphore = semaphore;
+    impSemInfo.handleType = VK_EXTERNAL_SEMAPHORE_HANDLE_TYPE_OPAQUE_FD_BIT;
+    impSemInfo.fd = semaphoreFd;
+
+    VK_CHECK(d.vkImportSemaphoreFdKHR(m_dev, &impSemInfo));
+#endif
+
+    VkImageViewCreateInfo viewInfo = {};
+    viewInfo.sType = VK_STRUCTURE_TYPE_IMAGE_VIEW_CREATE_INFO;
+    viewInfo.viewType = VK_IMAGE_VIEW_TYPE_2D;
+    viewInfo.format = imageInfo.format;
+    viewInfo.image = image;
+    viewInfo.subresourceRange = {};
+    viewInfo.subresourceRange.aspectMask = VK_IMAGE_ASPECT_COLOR_BIT;
+    viewInfo.subresourceRange.baseMipLevel = 0;
+    viewInfo.subresourceRange.levelCount = 1;
+    viewInfo.subresourceRange.baseArrayLayer = 0;
+    viewInfo.subresourceRange.layerCount = 1;
+    viewInfo.components.r = VK_COMPONENT_SWIZZLE_IDENTITY;
+    viewInfo.components.g = VK_COMPONENT_SWIZZLE_IDENTITY;
+    viewInfo.components.b = VK_COMPONENT_SWIZZLE_IDENTITY;
+    viewInfo.components.a = VK_COMPONENT_SWIZZLE_IDENTITY;
+    VkImageView view;
+    VK_CHECK(vkCreateImageView(m_dev, &viewInfo, nullptr, &view));
+
+    VkDescriptorSetAllocateInfo descriptorAllocInfo = {};
+    descriptorAllocInfo.sType = VK_STRUCTURE_TYPE_DESCRIPTOR_SET_ALLOCATE_INFO;
+    descriptorAllocInfo.descriptorSetCount = 1;
+    descriptorAllocInfo.pSetLayouts = &m_descriptorLayout;
+    descriptorAllocInfo.descriptorPool = m_descriptorPool;
+    VkDescriptorSet descriptor;
+    VK_CHECK(vkAllocateDescriptorSets(m_dev, &descriptorAllocInfo, &descriptor));
+
+    VkDescriptorImageInfo descriptorImageInfo = {};
+    descriptorImageInfo.imageView = view;
+    descriptorImageInfo.imageLayout = VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL;
+
+    VkWriteDescriptorSet descriptorWriteSet = {};
+    descriptorWriteSet.sType = VK_STRUCTURE_TYPE_WRITE_DESCRIPTOR_SET;
+    descriptorWriteSet.descriptorCount = 1;
+    descriptorWriteSet.descriptorType = VK_DESCRIPTOR_TYPE_COMBINED_IMAGE_SAMPLER;
+    descriptorWriteSet.dstSet = descriptor;
+    descriptorWriteSet.pImageInfo = &descriptorImageInfo;
+    vkUpdateDescriptorSets(m_dev, 1, &descriptorWriteSet, 0, nullptr);
+
+    VkImageMemoryBarrier imageBarrier = {};
+    imageBarrier.sType = VK_STRUCTURE_TYPE_IMAGE_MEMORY_BARRIER;
+    imageBarrier.oldLayout = VK_IMAGE_LAYOUT_UNDEFINED;
+    imageBarrier.newLayout = VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL;
+    imageBarrier.image = image;
+    imageBarrier.subresourceRange.aspectMask = VK_IMAGE_ASPECT_COLOR_BIT;
+    imageBarrier.subresourceRange.layerCount = 1;
+    imageBarrier.subresourceRange.levelCount = 1;
+    imageBarrier.srcAccessMask = VK_ACCESS_NONE;
+    imageBarrier.dstAccessMask = VK_ACCESS_SHADER_READ_BIT;
+    vkCmdPipelineBarrier(copyCmd, VK_PIPELINE_STAGE_BOTTOM_OF_PIPE_BIT, VK_PIPELINE_STAGE_FRAGMENT_SHADER_BIT, 0, 0, nullptr, 0, nullptr, 1, &imageBarrier);
```

Comment:
According to https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#synchronization-image-layout-transitions
> If the old layout is VK_IMAGE_LAYOUT_UNDEFINED, the contents of that range may be discarded.

You need to specify the actual layout, certainly `VK_IMAGE_LAYOUT_PRESENT_KHR`.

I don't know for multiprocess apps what actually happens, especially because synchronization is not really done.

Reply:
I'm not really sure about this.
The image was passed to AVVkFrame with `VK_IMAGE_LAYOUT_UNDEFINED` before too, and it worked.
No warnings from validation layer either.

Reply:
Does this apply for external memory images? Because the image may be in different layout at the compositor process (layer), but when it's created here it is in undefined layout initially.
Also it shouldn't even matter, because it's the memory that is shared, not the image itself.

So I think this transfer (undefined -> shader read) here is correct.

Reply:
I believe it applies for any image, let's use your code, and if someone notices tiling artifacts it is likely to come from this.